app_dart/lib/src/service/cache_service.dart

```diff
@@ -5,28 +5,85 @@
 import 'dart:convert';
 import 'dart:typed_data';
 
+import 'package:meta/meta.dart';
 import 'package:neat_cache/cache_provider.dart';
 import 'package:neat_cache/neat_cache.dart';
 
-import '../datastore/cocoon_config.dart';
-
 class CacheService {
-  CacheService(this.config);
-
-  final Config config;
-
-  CacheProvider<List<int>> _provider;
-
-  Future<Cache<Uint8List>> redisCache() async {
-    _provider = Cache.redisCacheProvider(await config.redisUrl);
-    return Cache<List<int>>(_provider)
-        .withCodec<Uint8List>(const _CacheCodec());
+  CacheService({
+    bool inMemory = false,
+    int inMemoryMaxSize = 256,
+  }) : _provider = inMemory
+            ? Cache.inMemoryCacheProvider(inMemoryMaxSize)
+            : Cache.redisCacheProvider(memorystoreUrl);
+
+  final CacheProvider<List<int>> _provider;
+
+  Cache<Uint8List> get cache =>
+      cacheValue ??
+      Cache<List<int>>(_provider).withCodec<Uint8List>(const _CacheCodec());
+
+  @visibleForTesting
+  Cache<Uint8List> cacheValue;
+
+  /// Google Cloud Memorystore default url.
+  static const String memorystoreUrl = 'redis://10.0.0.4:6379';
+
+  /// An arbritary number for how many times we should try to get from cache
+  /// before giving up.
+  ///
+  /// Writing to the cache creates a racy condition for when another operation
+  /// is trying to get the same key. This race condition throws an exception.
+  @visibleForTesting
+  static const int maxCacheGetAttempts = 3;
+
+  /// Get value of [key] from the subcache [subcacheName]. If the key has no
+  /// value, call [createFn] to create a value for it, set it, and return it.
+  ///
+  /// The underlying cache get function is inherently racy as if there is a
+  /// write operation while a read operation, getting the value can fail. To
+  /// handle this racy condition, this attempts to get the value [maxCacheGetAttempts]
+  /// times before giving up. This is because the cache is magnitudes faster
+  /// than the fallback operation (usually a Datastore query).
+  Future<Uint8List> get(
```

Comment:
Nit: since this function is doing two things, I'd call it `getOrCreate()`.

Reply:
Good catch! Added.